### DISCUSSION TITLE
Do cffi, docutils & pycparser depend on openssl@1.1?

### DISCUSSION THREAD

**Discussion by xyzzy-plugh-plover** [Homebrew/discussions#4784] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.1.11-2-geacdf7a
ORIGIN: https://github.com/Homebrew/brew.git
HEAD: eacdf7aa60a1a8a64acc71853ec5559cfd2329c3
Last commit: 9 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: fe20814d11c50c95df0bb6fb67f1f9ead05f7be6
Core tap last commit: 81 minutes ago
Core tap branch: master
Core tap JSON: 12 Sep 12:31 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.Rz7KNDZeqP/org.xquartz:0
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 4
HOMEBREW_NO_ANALYTICS: set
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: quad-core 64-bit haswell
Clang: 14.0.0 build 1400
Git: 2.42.0 => /usr/local/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 12.6.9-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  cffi
```

### Description of issue

`brew uses --installed openssl@1.1` returns `cffi`, `docutils` & `pycparser`, but `openssl` is not output by `brew info --verbose` for those casks, nor is it in their respective formulae.

Is `brew uses` returning incorrect info? If so, why, and how can I fix it? If it's correct, where are these `openssl@1.1` dependencies specified?

All of my formulae & casks are current.

**Comment by SMillerDev** [comment]
What does `brew outdated` say? Because AFAIK `uses` analyses your current install and `info` outputs general information.

**Reply by xyzzy-plugh-plover** [reply]
`brew outdated` says nothing.

All of my formulae & casks are current.

Where in my current install would it specify that the formulae depend on `openssl@1.1`?

Thanks for any help.

**Reply by gromgit** [reply]
To be sure, what's the output of `brew info cffi docutils pycparser`?

Note that the current versions of all those formulae depend indirectly on `openssl@3`, not `openssl@1.1`, and `brew info` only lists _direct_ dependencies. You need to run `brew deps` to get the full dependency picture.

**Reply by xyzzy-plugh-plover** [reply]
```
$ brew info cffi docutils pycparser
==> cffi: stable 1.15.1 (bottled)
C Foreign Function Interface for Python
https://cffi.readthedocs.io/en/latest/
/usr/local/Cellar/cffi/1.15.1 (33 files, 581.7KB)
  Poured from bottle using the formulae.brew.sh API on 2023-03-31 at 23:18:08
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/c/cffi.rb
License: MIT
==> Dependencies
Required: pycparser ✔, python@3.11 ✔

==> docutils: stable 0.20.1 (bottled)
Text processing system for reStructuredText
https://docutils.sourceforge.io
/usr/local/Cellar/docutils/0.20.1 (231 files, 2MB) *
  Poured from bottle using the formulae.brew.sh API on 2023-05-17 at 10:56:37
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/d/docutils.rb
License: Public Domain and BSD-2-Clause and GPL-3.0-or-later and Python-2.0
==> Dependencies
Required: python@3.11 ✔

==> pycparser: stable 2.21 (bottled)
C parser in Python
https://github.com/eliben/pycparser
/usr/local/Cellar/pycparser/2.21 (46 files, 656.7KB) *
  Poured from bottle using the formulae.brew.sh API on 2023-03-18 at 18:11:32
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/p/pycparser.rb
License: BSD-3-Clause
==> Dependencies
Required: python@3.11 ✔
```

**Reply by xyzzy-plugh-plover** [reply]
`python@3.11 ` depends on `openssl@3`. Somehow, `brew deps pycparser` changes that dependency to `openssl@1.1`.

```
$ brew deps pycparser
ca-certificates
mpdecimal
openssl@1.1
python@3.11
readline
sqlite
xz

$ brew deps --direct pycparser
python@3.11

$ brew deps python@3.11 
ca-certificates
mpdecimal
openssl@3
readline
sqlite
xz
```

**Reply by gromgit** [reply]
That's certainly odd:
```
x86_64 bigsur$ brew deps pycparser  
ca-certificates
mpdecimal
openssl@3
python@3.11
readline
sqlite
xz
```
What's the output of `cat $(brew --prefix pycparser)/INSTALL_RECEIPT.json`?

**Reply by xyzzy-plugh-plover** [reply]
```
$ cat $(brew --prefix pycparser)/INSTALL_RECEIPT.json
```

```json
{
  "homebrew_version": "4.0.4-86-g7c15dce",
  "used_options": [

  ],
  "unused_options": [

  ],
  "built_as_bottle": true,
  "poured_from_bottle": true,
  "loaded_from_api": true,
  "installed_as_dependency": true,
  "installed_on_request": false,
  "changed_files": [

  ],
  "time": 1679177492,
  "source_modified_time": 1636202539,
  "compiler": "clang",
  "aliases": [

  ],
  "runtime_dependencies": [
    {
      "full_name": "mpdecimal",
      "version": "2.5.1",
      "declared_directly": false
    },
    {
      "full_name": "ca-certificates",
      "version": "2023-01-10",
      "declared_directly": false
    },
    {
      "full_name": "openssl@1.1",
      "version": "1.1.1t",
      "declared_directly": false
    },
    {
      "full_name": "readline",
      "version": "8.2.1",
      "declared_directly": false
    },
    {
      "full_name": "sqlite",
      "version": "3.41.1",
      "declared_directly": false
    },
    {
      "full_name": "xz",
      "version": "5.4.1",
      "declared_directly": false
    },
    {
      "full_name": "python@3.11",
      "version": "3.11.2",
      "declared_directly": true
    }
  ],
  "source": {
    "spec": "stable",
    "versions": {
      "stable": "2.21",
      "head": null,
      "version_scheme": 0
    },
    "path": "/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/pycparser.rb",
    "tap_git_head": "d5af7df1eea4a09689297fd9038e110187bb7ca8",
    "tap": "homebrew/core"
  },
  "arch": "x86_64",
  "built_on": {
    "os": "Macintosh",
    "os_version": "macOS 13",
    "cpu_family": "dunno",
    "xcode": "14.2",
    "clt": "14.2.0.0.1.1668646533",
    "preferred_perl": "5.30"
  }
}
```

**Reply by gromgit** [reply]
Yeah, that `openssl@1.1` dependency is definitely not right, but I've no idea how that got there. Perhaps @SMillerDev could shed further light on this?

UPDATE: Never mind, just noticed that `python@3.11` was depending on `openssl@1.1` until 3.11.4, which postdates your `pycparser` installation. That's where it picked up that old dependency from.

**Reply by xyzzy-plugh-plover** [reply]
`brew deps` & `brew uses` shouldn't obtain recursive dependencies from installation receipts, as the recursive dependencies that they contain might be out of date; they should recursively request each formula's current direct dependencies. If the brew commands are using installation receipts for recursive dependencies (instead of solely for direct dependencies), then they are broken.

If any direct dependency is not installed, then it should be listed with an asterisk (or similar notation) to indicate that it is missing. The recursive dependencies from its direct parent's install receipt that aren't already listed in the output can be appended, but should have two asterisks (or similar notation) to indicate that they are recursive dependencies that are assumed to have been required by the missing direct dependency or its recursive dependencies.

**Reply by xyzzy-plugh-plover** [reply]
I've reinstalled the 3 formulae; the install receipts now include `openssl@3`, and both `brew deps` & `brew uses` show that the formulae use `openssl@3`. It is likely that the commands use the install receipts.

**Reply by xyzzy-plugh-plover** [reply]
Is the above the intended behavior?

If so, then the documentation should be updated to note that `brew deps` & `brew uses` use outdated information from installation receipts. (even if this is the intended behavior, it seems a poor choice of behavior)

If not, then the code should be fixed.

**Reply by cho-m** [reply]
Sounds similar to a known issue in `brew` (e.g. https://github.com/Homebrew/brew/issues/13717) where dependency information is based on formulae/versions at original install time.

It is probably not intended, but the behavior will remain until someone has time to work on a fix. Main workaround is to refresh the tab by `brew reinstall`.
